### PR TITLE
[3875][ADD] api_connection_key_encryption

### PR DESCRIPTION
**Pull request by AungKoKoLin1997**
[3875](https://www.quartile.co/web#id=3875&menu_id=505&cids=3&action=1457&model=project.task&view_type=form)

### CODE REVIEW

File: api_connection_key_encryption/README.rst [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+Authors
+-------
+
+* Quartile Limited
+
+Maintainers
+-----------
+
+This module is part of the `qrtl/axls-private <https://github.com/qrtl/axls-private/tree/16.0/api_connection_key_encryption>`_ project on GitHub.
```
Comment (kanda999) [inline api_connection_key_encryption/README.rst]:
```suggestion
This module is part of the `qrtl/axls-oca <https://github.com/qrtl/axls-oca/tree/16.0/api_connection_key_encryption>`_ project on GitHub.
```
We need to follow up on this.

**Review by kanda999 — APPROVED** [review]
Code review

File: api_connection_key_encryption/models/api_config.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+class ApiConfig(models.Model):
+    _inherit = "api.config"
+
+    encrypted_data_ids = fields.One2many("encrypted.data", "api_config_id")
+
+    def _current_env_encrypted_key_exists(self):
+        env = self.env["encrypted.data"]._retrieve_env()
+        key_name = "encryption_key_%s" % env
+        return bool(config.get(key_name))
```
Comment (yostashiro) [inline api_connection_key_encryption/models/api_config.py]:
```suggestion
    def _get_env_with_encryption_key(self):
        env = self.env["encrypted.data"]._retrieve_env()
        key_name = "encryption_key_%s" % env
        if not bool(config.get(key_name)):
            raise UserError(_("Encryption key not found for the current environment."))
        return env
```

File: api_connection_key_encryption/models/api_config.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    @api.model
+    def create(self, vals):
+        if not self._current_env_encrypted_key_exists():
+            raise UserError(_("Encryption key not found for the current environment."))
+        if "api_key" in vals and vals["api_key"]:
+            env = self.env["encrypted.data"]._retrieve_env()
```
Comment (yostashiro) [inline api_connection_key_encryption/models/api_config.py]:
```suggestion
        if "api_key" in vals and vals["api_key"]:
            env = self._get_env_with_encryption_key()
```

File: api_connection_key_encryption/models/api_config.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            rec.sudo().encrypted_data_ids.write({"encrypted_data": encrypted_key})
+        return super().write(vals)
+
+    def get_decrypted_api_key(self):
+        self.ensure_one()
+        if self._current_env_encrypted_key_exists():
+            env = self.env["encrypted.data"]._retrieve_env()
+            return self.sudo().encrypted_data_ids._decrypt_data(env)
+        else:
+            raise UserError(_("Encryption key not found for the current environment."))
```
Comment (yostashiro) [inline api_connection_key_encryption/models/api_config.py]:
```suggestion
        env = self._get_env_with_encryption_key()
        return self.sudo().encrypted_data_ids._decrypt_data(env)
```

File: api_connection_key_encryption/models/api_config.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                "environment": env,
+                "encrypted_data": encrypted_data,
+            }
+        )
+        return res
+
+    def write(self, vals):
+        if "api_key" not in vals:
+            return super().write(vals)
+        env = self.env["encrypted.data"]._retrieve_env()
```
Comment (yostashiro) [inline api_connection_key_encryption/models/api_config.py]:
```suggestion
        env = self._get_env_with_encryption_key()
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

**Review by yostashiro — APPROVED** [review]
Code review.

**Pull request merged**